server/src/main/java/org/eclipse/openvsx/mirror/MirrorExtensionService.java

```diff
@@ -0,0 +1,211 @@
+        var download = json.files.get("download");
+        var userJson = new UserJson();
+        userJson.provider = json.publishedBy.provider;
+        userJson.loginName = json.publishedBy.loginName;
+        userJson.fullName = json.publishedBy.fullName;
+        userJson.avatarUrl = json.publishedBy.avatarUrl;
+        userJson.homepage = json.publishedBy.homepage;
+        var namespaceName = json.namespace;
+
+        var vsixResourceHeaders = backgroundNonRedirectingRestTemplate.headForHeaders("{resolveVsixLocation}", Map.of("resolveVsixLocation", download));
+        var vsixLocation = vsixResourceHeaders.getLocation();
+        if (vsixLocation == null) {
+            throw new RuntimeException("Failed to parse location header from redirected vsix url");
+        }
+
+        var tokens = vsixLocation.getPath().split("/");
+        var filename = tokens[tokens.length-1];
+        if (!filename.endsWith(".vsix")) {
+            throw new RuntimeException("Invalid vsix filename from redirected vsix url");
+        }
+
+        URLConnection vsixConnection;
+        try {
+            vsixConnection = vsixLocation.toURL().openConnection();
+            vsixConnection.setConnectTimeout((int) Duration.ofSeconds(30).toMillis());
+            vsixConnection.setReadTimeout((int) Duration.ofMinutes(5).toMillis());
+        } catch (IOException e) {
+            throw new RuntimeException("failed to download vsix", e);
+        }
+        Path downloadPath;
+        try {
+            downloadPath = Files.createTempFile("extension_", ".vsix");
+        } catch (IOException e) {
+            throw new RuntimeException("failed to download vsix", e);
+        }
+        try {
+            try (
+                var vsisStream = vsixConnection.getInputStream();
+                var vsixChannel = Channels.newChannel(vsisStream);
+                var downlodStream =  new FileOutputStream(downloadPath.toFile());
+                var downlodChannel = downlodStream.getChannel();
+            ) {
+                downlodChannel.transferFrom(vsixChannel, 0, Long.MAX_VALUE);
+            } catch (IOException e) {
+                throw new RuntimeException("failed to download vsix", e);
+            }
+
+            var user = data.getOrAddUser(userJson);
+            var namespace = repositories.findNamespace(namespaceName);
+            data.ensureNamespaceMembership(user, namespace);
+
+            var description = "MirrorExtensionVersion";
+            var accessTokenValue = data.getOrAddAccessTokenValue(user, description);
+
+            var token = users.useAccessToken(accessTokenValue);
+            extensions.mirrorVersion(downloadPath, token, filename, json.timestamp);
+            LOGGER.debug("completed mirroring of extension version: {}", json.namespace + "." + json.name + "-" + json.version + "@" + json.targetPlatform);
+        } finally {
+            try {
+                Files.delete(downloadPath);
```

Comment:
Not really necessary

Reply:
https://github.com/eclipse/openvsx/pull/586#discussion_r1036826240